SharpToken/Lib/Encoding.cs

```diff
@@ -126,14 +136,44 @@ public string Decode(List<int> inputTokensToDecode)
             return Encoding.UTF8.GetString(decodedBytes);
         }
 
-        private static int GetMaxValueFromDictionary(Dictionary<byte[], int> dictionary)
-        {
-            return dictionary.Values.Prepend(0).Max();
-        }
 
-        private static int GetMaxValueFromDictionary(Dictionary<string, int> dictionary)
+        #region Private
+
+        private (List<int> tokens, int count) EncodeCore(
+#if NET8_0_OR_GREATER
+            ReadOnlySpan<char> lineToEncode,
+#else
+            string lineToEncode,
+#endif
+            ISet<string> allowedSpecial = null,
+            ISet<string> disallowedSpecial = null,
+            bool countOnly = false
+        )
         {
-            return dictionary.Values.Prepend(0).Max();
+            var allowedSpecialTokens = allowedSpecial is null
+                // When null allow nothing
+                ? Array.Empty<string>()
```

Comment:
Line: 155: `Array.Empty<string>()` is important because `allowedSpecialTokens.FindMatch()` implements a no-allocation fast path for it.